### PR TITLE
improve `Selection::append_selection` and `Selection::replace_with_selection`

### PR DESCRIPTION
**Pull request by niklak**
## Summary by CodeRabbit

- **New Features**
  - Improved performance of `Selection::append_selection` and `Selection::replace_with_selection` by optimizing internal logic.

- **Tests**
  - Expanded the test suite for selection manipulation, covering various scenarios and ensuring robust functionality, including edge cases with empty selections.

### CODE REVIEW

File: src/selection.rs [hunk 1]
```diff
@@ -435,17 +434,17 @@
...
-        sel.iter().for_each(|s| contents.push_tendril(&s.html()));
-        let fragment = Document::from(contents);
         sel.remove();
 
+        let sel_nodes = sel.nodes();
         for node in self.nodes() {
-            node.tree.merge_with_fn(fragment.tree.clone(), |node_id| {
-                node.append_prev_siblings(&node_id)
+            node.tree.copy_nodes_with_fn(sel_nodes, |new_node_id| {
+                node.append_prev_sibling(&new_node_id)
```
[the remark this answers is not in the document]
Comment (niklak) [reply src/selection.rs:447]: It is a feature to support cross-tree operations...

File: src/dom_tree.rs [hunk 2]
```diff
@@ -519,4 +520,91 @@
...
+    /// * `f` - function to be applied to the first copied node
+    pub(crate) fn copy_nodes_with_fn<F>(&self, other_nodes: &[NodeRef], f: F)
+    where
+        F: Fn(NodeId),
+    {
+        for other_node in other_nodes {
+            let new_node_id = self.copy_node(other_node);
+            f(new_node_id);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (niklak) [reply src/dom_tree.rs:609]: You are very wrong. The function applies to the first (parent) node on the iteration. So it copies a node and its childdren and then applies a function only for the first node.

File: src/dom_tree.rs [hunk 3]
```diff
@@ -519,4 +520,91 @@ impl Tree {
...
+
+        // source tree may be the same tree that owns the copy_node fn, and may be not.
+        let source_tree = node.tree;
+        let new_nodes = self.copy_tree_nodes(source_tree, &id_map);
+
+        let mut nodes = self.nodes.borrow_mut();
+        nodes.extend(new_nodes);
+
+        base_id
+    }
```
[the remark this answers is not in the document]
Comment (niklak) [reply src/dom_tree.rs:559]: `Tree::copy_node` and `Tree::copy_nodes_with_fn` are both crate internal methods. So they will not appear to doc.rs and will be unavailable for the user. I know that `Tree` using RefCell, but it wasn`t my choice, it is a consequence of html5ever crate update, they`ve change the semantics of TreeSink trait.
